network/p2p/inspector/internal/cache/cache.go

```diff
@@ -0,0 +1,229 @@
+// Returns:
+// - *RecordCache, the created cache.
+// Note that this cache is supposed to keep the cluster prefix control messages received record for the authorized (staked) nodes. Since the number of such nodes is
+// expected to be small, we do not eject any records from the cache. The cache size must be large enough to hold all
+// the records of the authorized nodes. Also, this cache is keeping at most one record per peer id, so the
+// size of the cache must be at least the number of authorized nodes.
+func NewRecordCache(config *RecordCacheConfig, recordEntityFactory recordEntityFactory) (*RecordCache, error) {
+	backData := herocache.NewCache(config.sizeLimit,
+		herocache.DefaultOversizeFactor,
+		// this cache is supposed to keep the cluster prefix control messages received record for the authorized (staked) nodes. Since the number of such nodes is
+		// expected to be small, we do not eject any records from the cache. The cache size must be large enough to hold all
+		// the records of the authorized nodes. Also, this cache is keeping at most one record per peer id, so the
+		// size of the cache must be at least the number of authorized nodes.
+		heropool.NoEjection,
+		config.logger.With().Str("mempool", "gossipsub=cluster-prefix-control-messages-received-records").Logger(),
+		config.collector)
+	return &RecordCache{
+		recordEntityFactory: recordEntityFactory,
+		decayFunc:           defaultDecayFunction(config.recordDecay),
+		c:                   stdmap.NewBackend(stdmap.WithBackData(backData)),
+	}, nil
+}
+
+// Init initializes the record cache for the given peer id if it does not exist.
+// Returns true if the record is initialized, false otherwise (i.e.: the record already exists).
+// Args:
+// - nodeID: the node ID of the sender of the control message.
+// Returns:
+// - true if the record is initialized, false otherwise (i.e.: the record already exists).
+// Note that if Init is called multiple times for the same peer id, the record is initialized only once, and the
+// subsequent calls return false and do not change the record (i.e.: the record is not re-initialized).
+func (r *RecordCache) Init(nodeID flow.Identifier) bool {
+	entity := r.recordEntityFactory(nodeID)
+	return r.c.Add(entity)
+}
+
+// Update applies an adjustment that increments the number of cluster prefixed control messages received by a peer.
+// Returns number of cluster prefix control messages received after the adjustment. The record is initialized before
+// the adjustment func is applied that will increment the Gauge.
+// Args:
+// - nodeID: the node ID of the sender of the control message.
+// - adjustFunc: the function that adjusts the record.
+// Returns:
+//   - The cluster prefix control messages received gauge value after the adjustment.
+//   - exception only in cases of internal data inconsistency or bugs. No errors are expected.
+func (r *RecordCache) Update(nodeID flow.Identifier) (float64, error) {
```

Comment:
```suggestion
func (r *RecordCache) ReceivedSpam(nodeID flow.Identifier) (float64, error) {
```
This module is very use-case-specific -- I feel we could have a more descriptive name than `Update`, to reflect that it is incrementing a counter, that it corresponds to receiving one "spam" message,  etc.

Reply:
https://github.com/onflow/flow-go/pull/4365/commits/ac998fe58e12c603b56b53771ea83f2ce512d999